rdf.1
.TH rdf.sh 1  "10/2017" "version 0.8.0" "USER COMMANDS"
.SH NAME
rdf.sh \- A multi-tool shell script for doing Semantic Web jobs on the command line.
.SH SYNOPSIS
.B rdf
[subcommand]
[option]
.SH DESCRIPTION
The idea of this tool is to provide a fast command line interface to day by day Semantic Web / RDF hacker related jobs.
.PP
This includes inspecting resources and namespaces, listing of schema elements and other tasks.
.SS subcommands
    color           -- get a html color for a resource URI
    count           -- count distinct triples
    delete          -- deletes an existing linked data resource via LDP
    desc            -- outputs description of the given resource in a given format (default: turtle)
    diff            -- diff of triples from two RDF files
    edit            -- edit the content of an existing linked data resource via LDP (GET + PUT)
    get             -- curls rdf in xml or turtle to stdout (tries accept header)
    get-ntriples    -- curls rdf and transforms to ntriples
    head            -- curls only the http header but accepts only rdf
    headn           -- curls only the http header
    help            -- outputs the manpage of rdf
    list            -- list resources which start with the given URI
    ns              -- curls the namespace from prefix.cc
    nscollect       -- collects prefix declarations of a list of ttl/n3 files
    nsdist          -- distributes prefix declarations from one file to a list of other ttl/n3 files
    nssort          -- sorts the prefix declarations in files, inserting all at the line of the first found prefix
    ping            -- sends a semantic pingback request from a source to a target or to all possible targets
    pingall         -- sends a semantic pingback request to all possible targets of a given resource
    put             -- replaces an existing linked data resource via LDP
    split           -- split an RDF file into pieces of max X triple and -optional- run a command on each part
    turtleize       -- outputs an RDF file in turtle, using as much as possible prefix declarations
.TP
Each subcommand outputs a detailed help by using it without parameter:

.B $ rdf count
    Syntax: rdf count <file>
    (count triples using rapper)

.SH EXAMPLE
.SS namespace lookup (ns)
.TP
rdf.sh allows you to quickly lookup namespaces from prefix.cc:

.B $ rdf ns foaf
    http://xmlns.com/foaf/0.1/
.TP
rdf.sh can also output prefix.cc syntax templates: 
.B $ rdf ns skos sparql
    PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

    SELECT *
    WHERE {
      ?s ?p ?o .
    }

.B $ rdf ns ping n3
    @prefix ping: <http://purl.org/net/pingback/> .
.TP
Note: As a result of this subcommand, all other rdf command can get qnames as parameters (e.g. foaf:Person or skos:Concept) which results in a namespace lookup ahead of the used command.
.SS resource description (desc)
.TP
Describe a resource by querying for statements where the resource is the subject. This is extremly useful to fastly check schema details.

.B $ rdf desc foaf:Person

.B $ rdf desc skos:narrower

.SS resource listings (list)
.TP
To get a quick overview of an unknown RDF schema, rdf.sh provides the `list` command which outputs a distinct list of subject resources of the fetched URI:

.B $ rdf list geo:
    http://www.w3.org/2003/01/geo/wgs84_pos#
    http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing
    http://www.w3.org/2003/01/geo/wgs84_pos#Point
    http://www.w3.org/2003/01/geo/wgs84_pos#lat
    http://www.w3.org/2003/01/geo/wgs84_pos#location
    http://www.w3.org/2003/01/geo/wgs84_pos#long
    http://www.w3.org/2003/01/geo/wgs84_pos#alt
    http://www.w3.org/2003/01/geo/wgs84_pos#lat_long
.TP
You can also provide a starting sequence to constrain the output

.B $ rdf list skos:C
    http://www.w3.org/2004/02/skos/core#Concept
    http://www.w3.org/2004/02/skos/core#ConceptScheme
    http://www.w3.org/2004/02/skos/core#Collection
    http://www.w3.org/2004/02/skos/core#changeNote
    http://www.w3.org/2004/02/skos/core#closeMatch
.TP
Note: Here the $GREP_OPTIONS environment applies to the list. In my case, I have a --ignore-case in it, so e.g. skos:changeNote is listed as well.
.TP
This feature only works with schema documents which are available by fetching the namespace URI (optionally with linked data headers to be redirected to an RDF document). Nevertheless, you can use this command also on non schema resources as FOAF profiles and WebIDs:

.B $ rdf list http://haschek.eye48.com/
    http://haschek.eye48.com/haschek.rdf
    http://haschek.eye48.com/
    http://haschek.eye48.com/gelabb/
.SS resources inspection (get, count, head and headn)

.TP
Fetch a resource via linked data and print it to stdout:

.B $ rdf get http://sebastian.tramp.name >me.rdf
.TP
Count all statements of a resource (using rapper):

.B $ rdf count http://sebastian.tramp.name
    rapper: Parsing URI http://sebastian.tramp.name with parser guess
    rapper: Parsing returned 58 triples
.TP
Inspect the header of a resource. Use `head` for header request with content negotiation suitable for linked data and `headn` for a normal header request as sent by browsers.

.B $ rdf head http://sebastian.tramp.name
    HTTP/1.1 302 Found
    [...]
    Location: http://sebastian.tramp.name/index.rdf
    [...]

.B $ rdf ping http://sebastian.tramp.name http://aksw.org/SebastianTramp
    server response: Pingback has been registered or updated... Keep spinning the Data Web ;-)

.SH INSTALLATION
.TP
rdf.sh is a single bash shell script so installation is trivial ... :-) Just copy or link it to you path, e.g. with

.B $ sudo ln -s /path/to/rdf.sh /usr/local/bin/rdf
.TP 
Required tools currently are:
  * roqet -- http://librdf.org/rasqal/roqet.html (from rasqal-utils)
  * rapper -- http://librdf.org/raptor/rapper.html (from raptor-utils or raptor2-utils)
  * curl -- http://curl.haxx.se/
.TP
Suggested tools are:
  * zsh -- http://zsh.sourceforge.net/ (without the autocompletion, it is not the same)

.SH FILES
.B changelog.md
version changelog

.B _rdf
zsh autocompletion file

.B rdf.sh
the script

.B rdf.1
this file

.B $HOME/.cache/rdf.sh/prefix.cache
cached namespace lookup table from prefix.cc

.B $HOME/.config/rdf.sh/prefix.local
user-generated local priority lookup table

.SH AUTHOR
.SS Sebastian Tramp <mail@sebastian.tramp.name>
.TP
http://sebastian.tramp.name (WebID)
.SH SEE ALSO
.SS https://github.com/seebi/rdf.sh
.TP
The main homepage at github



ChangeLog
0.1
- first version
- available commands: get head rdfhead ns diff count desc list split


rdf.sh
#!/usr/bin/env bash
# @(#) A multi-tool shell script for doing Semantic Web jobs on the command line.

# application metadata
name="rdf.sh"
version="0.5"
home="https://github.com/seebi/rdf.sh"

# basic application environment
this=`basename $0`
thisexec=$0
command="$1"
curlcommand="curl --fail -A ${name}/${version} -s -L"
os=`uname -s | tr "[:upper:]" "[:lower:]"`

if [ "$RDFSH_ACCEPT_HEADER" == "" ]
then
    mimetypes="text/turtle; q=1.0, application/x-turtle; q=0.9, text/n3; q=0.8, application/rdf+xml; q=0.5, text/plain; q=0.1"
else
    mimetypes="$RDFSH_ACCEPT_HEADER"
    echo "Will use '$mimetypes' for content negotiation."
fi

### mac workarounds
uname=`uname`
if [ "$uname" == "Darwin" ]
then
    sed="sed -E"
    sedi="sed -i -E"
else
    sed="sed"
    sedi="sed -i"
fi


###
# private functions
###

# takes an input command name and checks for availability with which
_checkTool ()
{
    tool=$1
    if [ "$tool" == "" ]
    then
        echo "checkTool error: need an parameter"
        exit 1
    fi

    for tool in $*
    do
        check=`which $tool 2>/dev/null >/dev/null`
        if [ "$?" == "1" ]
        then
            echo "Error: you need '$tool' for this command."
            exit 1
        fi
    done
}

# takes an input string and checks if it is a valid qname
_isQName ()
{
    _checkTool cut
    qname=$1
    if [ "$qname" == "" ]
    then
        echo "isQName error: need an parameter"
        exit 1
    fi

    LocalPart=`echo $qname | cut -d ":" -f 2`
    if [ "$qname" == "$LocalPart" ]
    then
        echo "false"
        return
    else
        Prefix=`echo $qname | cut -d ":" -f 1`

        # this is ugly ... here we distinguish between uris and qnames
        case "$Prefix" in
            "http" | "https" | "mailto" | "ldap" | "urn" )
                echo "false"
                return
            ;;
        esac

        if [ "$qname" != "$Prefix:$LocalPart" ]
        then
        echo "false"
        return
        else
            echo "true"
            return
        fi
    fi
}

# takes a qname and outputs the prefix
_getPrefix ()
{
    _checkTool cut
    qname=$1
    if [ "$qname" == "" ]
    then
        echo "getPrefix error: need an qname parameter"
        exit 1
    fi

    LocalPart=`echo $qname | cut -d ":" -f 2`
    if [ "$qname" == "$LocalPart" ]
    then
        echo "getPrefix error: $qname is not a valid qname"
        exit 1
    else
        Prefix=`echo $qname | cut -d ":" -f 1`
        if [ "$qname" != "$Prefix:$LocalPart" ]
        then
        echo "getPrefix error: $qname is not a valid qname"
        exit 1
        else
            echo $Prefix
        fi
    fi
}

# takes a qname and outputs the LocalName
_getLocalName ()
{
    _checkTool cut
    qname=$1
    if [ "$qname" == "" ]
    then
        echo "getLocalName error: need an qname parameter"
        exit 1
    fi

    LocalPart=`echo $qname | cut -d ":" -f 2`
    if [ "$qname" == "$LocalPart" ]
    then
        echo "getLocalName error: $qname is not a valid qname"
        exit 1
    else
        Prefix=`echo $qname | cut -d ":" -f 1`
        if [ "$qname" != "$Prefix:$LocalPart" ]
        then
        echo "getLocalName error: $qname is not a valid qname"
        exit 1
        else
            echo $LocalPart
        fi
    fi
}

# takes an input qname or URI and outputs the expanded full URI (if it is a qname)
_expandQName ()
{
    input=$1
    isQName=`_isQName $input`
    if [ "$isQName" == "true" ]
    then
        prefix=`_getPrefix $input`
        localName=`_getLocalName $input`
        namespace=`$thisexec ns $prefix`
        echo $namespace$localName
    else
        echo $input
    fi
}

_getNamespaceFromPrefix ()
{
    _checkTool curl cut
    prefix=$1
    if [ "$prefix" == "" ]
    then
        echo "getNamespaceFromPrefix error: need a prefix parameter"
        exit 1
    fi
    namespace=`_getNamespaceForPrefix $prefix`
    if [ "$namespace" == "" ]
    then
        # no cache-hit, request it from prefix.cc
        namespace=`$curlcommand http://prefix.cc/$prefix.file.n3 | cut -d "<" -f 2 | cut -d ">" -f 1`
        if [ "$namespace" != "" ]
        then
            _addPrefixToCache $prefix "$namespace"
        fi
    fi
    # output cache hit or curl output (maybe empty)
    echo $namespace
}

# give a namespace and get a prefix or it
# this function search in the cache as well the locally configured prefixes
_getPrefixForNamespace ()
{
    _checkTool cat grep head cut
    namespace=$1
    if [ "$namespace" == "" ]
    then
        echo "getPrefixFromCache error: need a namespace parameter"
        exit 1
    fi
    prefix=`cat $prefixlocal $prefixcache | grep "$namespace" | head -1 | cut -d "|" -f 1`
    echo $prefix
}

# give a prefix and get a namespace or ""
# this function search in the cache as well the locally configured prefixes
_getNamespaceForPrefix ()
{
    _checkTool cat grep head cut
    prefix=$1
    if [ "$prefix" == "" ]
    then
        echo "getPrefixFromCache error: need a prefix parameter"
        exit 1
    fi
    namespace=`cat $prefixlocal $prefixcache | grep "^$prefix|" | head -1 | cut -d "|" -f 2`
    echo $namespace
}

# give a prefix + namespace and get a new cache entry
_addPrefixToCache ()
{
    prefix=$1
    if [ "$prefix" == "" ]
    then
        echo "addPrefixToCache error: need a prefix parameter"
        exit 1
    fi
    namespace=$2
    if [ "$namespace" == "" ]
    then
        echo "addPrefixToCache error: need a namespace parameter"
        exit 1
    fi
    touch $prefixcache
    existingNamespace=`_getNamespaceForPrefix $prefix`
    if [ "$existingNamespace" == "" ]
    then
        echo "$prefix|$namespace" >>$prefixcache
    fi
}

# add a resource to the .resource_history file
_addToHistory ()
{
    _checkTool grep wc sed
    resource=$1
    if [ "$resource" == "" ]
    then
        echo "addToHistory error: need an resource parameter"
        exit 1
    fi

    historyfile=$2
    if [ "$historyfile" == "" ]
    then
        echo "addToHistory error: need an historyfile as second parameter "
        exit 1
    fi
    touch $historyfile

    if [ "$noHistory" == "" ]
    then
        count=`grep $resource $historyfile | wc -l`
        if [ "$count" != 0 ]
        then
            # f resource exists, remove it
            $sedi "s|$resource||g" $historyfile
            $sedi '/^$/d' $historyfile
        fi
        # add (or re-add) the resource at the end
        echo $resource >>$historyfile
    fi
}

# creates a tempfile and returns the filename
_getTempFile ()
{
    _checkTool mktemp
    tmpfile=`mktemp -q ./rdfsh-XXXX.tmp`
    echo $tmpfile
}

# get all related resources
_getRelatedResources ()
{
    _checkTool roqet cut grep rm
    resource=$1
    if [ "$resource" == "" ]
    then
        echo "getRelatedResources error: need an resource parameter"
        exit 1
    fi
    uri=`_expandQName $resource`

    tmpfile=`_getTempFile`
    $thisexec get-ntriples $uri >$tmpfile.nt
    roqet -q -e "SELECT DISTINCT ?o {<$uri> ?property ?o. FILTER(isUri(?o))}" -D $tmpfile.nt 2>/dev/null | cut -d "<" -f 2 | cut -d ">" -f 1 | grep "^http"
    rm $tmpfile $tmpfile.nt
}

# returns the announced pingback URL or an empty string
_isPingbackEnabled ()
{
    _checkTool roqet head cut grep
    resource=$1
    if [ "$resource" == "" ]
    then
        echo "isPingbackEnabled error: need an resource parameter"
        exit 1
    fi
    uri=`_expandQName $resource`

    pingbackServer=`$thisexec head $uri | grep X-Pingback: | cut -d " " -f 2`
    #$thisexec head $uri | grep X-Pingback: | cut -d " " -f 2
    if [ "$pingbackServer" == "" ]
    then
        tmpfile=`_getTempFile`
        $thisexec get-ntriples $uri >$tmpfile.nt
        pingbackServer=`roqet -q -e "SELECT ?o {<$uri> <http://purl.org/net/pingback/to> ?o}" -D $tmpfile.nt 2>/dev/null | head -1 | cut -d "<" -f 2 | cut -d ">" -f 1`
        rm $tmpfile $tmpfile.nt
    fi

    # output server
    if [ "$pingbackServer" != "" ]
    then
        echo $pingbackServer
    fi
}

# send a pingback to server $1 which connects the source $2 with the target $3
_sendPingback ()
{
    pbserver=$1
    if [ "$pbserver" == "" ]
    then
        echo "sendPingback error: need a pingback server URL as first parameter"
        exit 1
    fi
    pbsource=$2
    if [ "$pbsource" == "" ]
    then
        echo "sendPingback error: need a pingback source URL as second parameter"
        exit 1
    fi
    pbtarget=$3
    if [ "$pbtarget" == "" ]
    then
        echo "sendPingback error: need a pingback target URL as third parameter"
        exit 1
    fi

    result=`$curlcommand $pbserver --data "source=$pbsource&target=$pbtarget"`
    echo "$result"
}

###
# the "command" functions:
# the are executed by using the first parameter and get all parameters as options
###

docu_desc () { echo "outputs description of the given resource in a given format (default: turtle)";}
do_desc ()
{
    _checkTool mv cat grep cut wc roqet rapper rm
    uri="$2"
    output="$3"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart> <format>"
        echo "(`docu_desc`)"
        exit 1
    fi
    if [ "$output" == "" ]
    then
        output="turtle"
    fi
    uri=`_expandQName $uri`
    tmpfile=`_getTempFile`
    $thisexec get-ntriples $uri >$tmpfile.nt
    roqet -q -e "CONSTRUCT {<$uri> ?p ?o} WHERE {<$uri> ?p ?o}" -D $tmpfile.nt -r ntriples >$tmpfile.out

    # try to prepare the prefix definitions
    #features='-f xmlns:foaf="http://xmlns.com/foaf/0.1/" -f xmlns:site="http://ns.ontowiki.net/SysOnt/Site/"'
    features=""
    for namespace in `cat $prefixlocal $prefixcache| grep "|" | cut -d "|" -f 2`
    do
        namespaceCount=`grep -P "$namespace[a-zA-Z]+\>" $tmpfile.out | wc -l`
        if [[ "$namespaceCount" -ge 1 ]]; then
            prefix=`_getPrefixForNamespace $namespace`
            #echo "Found $namespace -> $prefix"
            features="$features -f xmlns:$prefix=\"$namespace\""
        fi
    done

    rapper -q ${features} -i ntriples $tmpfile.out -o $output
    rm $tmpfile $tmpfile.out $tmpfile.nt
    _addToHistory $uri $historyfile
}

docu_list () { echo "list resources which start with the given URI"; }
do_list ()
{
    _checkTool roqet cut grep rm
    uri="$2"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart>"
        echo "(`docu_list`)"
        exit 1
    fi
    uri=`_expandQName $uri`
    tmpfile=`_getTempFile`
    # turn history off for this internal call (dirty URIs)
    export noHistory="true"
    $thisexec get-ntriples $uri >$tmpfile.nt
    roqet -q -e "SELECT DISTINCT ?s WHERE {?s ?p ?o. FILTER isURI(?s) } " -D $tmpfile.nt 2>/dev/null | cut -d "<" -f 2 | cut -d ">" -f 1 | grep $uri
    rm $tmpfile $tmpfile.nt
}

docu_get () { echo "curls rdf in xml or turtle to stdout (tries accept header)"; }
do_get ()
{
    _checkTool curl
    uri="$2"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart>"
        echo "(`docu_get`)"
        exit 1
    fi
    uri=`_expandQName $uri`
    $curlcommand -H "Accept: $mimetypes" $uri
    _addToHistory $uri $historyfile
}

docu_get-ntriples () { echo "curls rdf and transforms to ntriples"; }
do_get-ntriples ()
{
    _checkTool curl rapper
    uri="$2"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart>"
        echo "(`docu_get-ntriples`)"
        exit 1
    fi
    uri=`_expandQName $uri`
    $curlcommand -H "Accept: application/rdf+xml,text/turtle,application/x-turtle" $uri | rapper -q -i guess -o ntriples -I $uri -
    _addToHistory $uri $historyfile
}

docu_headn () { echo "curls only the http header"; }
do_headn ()
{
    _checkTool curl
    uri="$2"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart>"
        echo "(`docu_head`)"
        exit 1
    fi
    uri=`_expandQName $uri`
    $curlcommand -I -X HEAD $uri
    _addToHistory $uri $historyfile
}

docu_head () { echo "curls only the http header but accepts only rdf"; }
do_head ()
{
    _checkTool curl
    uri="$2"
    if [ "$uri" == "" ]
    then
        echo "Syntax:" $this "$command <URI | Prefix:LocalPart>"
        echo "(`docu_rdfhead`)"
        exit 1
    fi
    uri=`_expandQName $uri`
    $curlcommand -I -X HEAD -H "Accept: $mimetypes" $uri
    _addToHistory $uri $historyfile
}

docu_ns () { echo "curls the namespace from prefix.cc"; }
do_ns ()
{
    _checkTool curl
    prefix="$2"
    suffix="$3"
    if [ "$prefix" == "" ]
    then
        echo "Syntax:" $this "$command <prefix> <suffix>"
        echo "(`docu_ns`)"
        echo " suffix can be n3, rdfa, sparql, ...)"
        exit 1
    fi
    if [ "$suffix" == "" ]
    then
        # this is a standard request as "rdf ns foaf"
        namespace=`_getNamespaceFromPrefix $prefix`
        echo $namespace
    else
        if [ "$suffix" == "plain" ]
        then
            # this is for vim integration, plain = without newline
            namespace=`_getNamespaceFromPrefix $prefix`
            echo -n $namespace
        else
            # if a real suffix is given, we always fetch from prefix.cc
            $curlcommand http://prefix.cc/$prefix.file.$suffix
        fi
    fi
}

docu_diff () { echo "diff of two RDF files"; }
do_diff ()
{
    _checkTool rapper rm sort
    source1="$2"
    source2="$3"
    difftool="$4"

    if [ "$difftool" != "" ]
    then
        RDFSHDIFF=$difftool
    else
        _checkTool diff
        RDFSHDIFF="diff"
    fi

    if [ "$source2" == "" ]
    then
        echo "Syntax:" $this "$command <rdf-file-1> <rdf-file-2>"
        echo "(`docu_diff`)"
        exit 1
    fi
    dest1="/tmp/$RANDOM-`basename $source1`"
    dest2="/tmp/$RANDOM-`basename $source2`"
    rapper -i guess $source1 | sort >$dest1
    rapper -i guess $source2 | sort >$dest2
    $RDFSHDIFF $dest1 $dest2
    rm $dest1 $dest2
}

docu_count () { echo "count triples using rapper"; }
do_count ()
{
    _checkTool rapper
    file="$2"
    if [ "$file" == "" ]
    then
        echo "Syntax:" $this "$command <file>"
        echo "(`docu_count`)"
        exit 1
    fi
    rapper -i guess --count $file
}

docu_split () { echo "split an RDF file into pieces of max X triple and -optional- run a command on each part"; }
do_split ()
{
    _checkTool rapper split wc sed
    file="$2"
    size="$3"
    todo="$4"
    if [ "$file" == "" ]
    then
        echo "Syntax:" $this "$command <file> <size-X> <command>"
        echo "(`docu_split`)"
        exit 1
    fi
    if [ "$size" == "" ]
    then
        size="25000"
    fi

    tmpdir=`mktemp -d`
    rapper -i guess -q $file | split -a 5 -l $size - $tmpdir/
    echo "Input splitted into `ls -1 $tmpdir | wc -l` pieces of max. $size triples."

    if [ "$todo" != "" ]
    then
        echo "Now executing '$todo' for every part (using %PART% as a placeholder):"
        for piece in `ls -1 $tmpdir`
        do
            cd $tmpdir
            realtodo=`echo $todo | $sed "s/%PART%/$piece/" `
            echo $realtodo
            $realtodo
        done
    fi

    echo "The pieces are in $tmpdir ... "
}

docu_nscollect() { echo "collects prefix declarations of a list of ttl/n3 files";}
do_nscollect()
{
    _checkTool cat wc grep sort
    prefixfile="$2"

    if [ "$prefixfile" == "" ]
    then
        prefixfile="prefixes.n3"
    fi

    if [ -f "$prefixfile" ]
    then
        countBefore=`cat $prefixfile| wc -l`
    else
        countBefore=0
    fi

    files=`ls -1 *.n3 *.ttl 2>/dev/null | grep -v $prefixfile`
    cat $files | grep "@prefix " | sort -u >$prefixfile
    count=`cat $prefixfile| wc -l`
    echo "$count prefixes collected in $prefixfile ($countBefore before)"
    #for n3file in $files
    #do
    #done
}

docu_nsdist () { echo "distributes prefix declarations from one file to a list of other ttl/n3 files";}
do_nsdist ()
{
    _checkTool grep mktemp wc
    prefixfile="prefixes.n3"
    if [ ! -f "$prefixfile" ]
    then
        echo "Syntax:" $this "$command <targetfiles>"
        echo "(`docu_nsdist`)"
        echo "I try to use $prefixfile as source but it is empty."
        exit 1
    fi

    if [ "$2" == "" ]
    then
        files=`ls -1 *.n3 *.ttl 2>/dev/null | grep -v $prefixfile`
    else
        files=$@
    fi

    count=`cat $prefixfile| wc -l`
    tmpfile=`mktemp -q ./rdfsh-XXXX`
    for target in $files
    do
        if [ -f "$target" ]
        then
            before=`cat $target | grep "@prefix "  | wc -l`

            cat $target | grep -v "@prefix " >$tmpfile
            cat $prefixfile >$target
            cat $tmpfile >>$target

            after=`cat $target | grep "@prefix "  | wc -l`
            let result=$after-$before
            if [ "$result" -ge "0" ]
            then
                echo "$target: +$result prefix declarations"
            else
                echo "$target: $result prefix declarations"
            fi
        fi
    done
    rm $tmpfile
}

docu_pingall () { echo "sends a semantic pingback request to all possible targets of a given resource"; }
do_pingall ()
{
    _checkTool head
    pingsource="$2"

    # check for ping source
    if [ "$pingsource" == "" ]
    then
        echo "Syntax:" $this "$command <pingsource>"
        echo "(`docu_ping`)"
        exit 1
    fi
    pingsource=`_expandQName $pingsource`

    # search for possible targets
    pingtargets=`_getRelatedResources $pingsource`
    if [ "$pingtargets" == "" ]
    then
        echo "Error: No targets available at $pingsource."
        exit 1
    fi

    for target in $pingtargets
    do
        # look for a pingback server responsable for the target
        pingserver=`_isPingbackEnabled $target`
        if [ "$pingserver" == "" ]
        then
            echo "$target: No pingback server found."
        else
            # finally, do the ping
            response=`_sendPingback $pingserver $pingsource $target | head -1`
            echo "$target (response): $response"
        fi
    done
}

docu_ping () { echo "sends a semantic pingback request from a source to a target or to all possible targets"; }
do_ping ()
{
    _checkTool grep wc
    pingsource="$2"
    pingtargets="$3"

    # check for ping source
    if [ "$pingsource" == "" ]
    then
        echo "Syntax:" $this "$command <pingsource> <pingtarget>"
        echo "(`docu_ping`)"
        exit 1
    fi
    pingsource=`_expandQName $pingsource`

    # check for ping target
    if [ "$pingtargets" != "" ]
    then
        # target is given as parameter, so check
        pingtargets=`_getRelatedResources $pingsource | grep "^$pingtargets$"`
        if [ "$pingtargets" == "" ]
        then
            echo "Error: No link found."
            exit 1
        fi
    else
        # target not given as paremeter, so search for
        pingtargets=`_getRelatedResources $pingsource`
        if [ "$pingtargets" == "" ]
        then
            echo "Error: No targets available at all."
            exit 1
        fi
    fi

    count=`echo $pingtargets | wc -w`
    if [ "$count" -ge 2 ]
    then
        echo "Please provide one of these ping target resources as a second parameter:"
        for target in $pingtargets
        do
            echo "- $target"
        done
        exit 1
    fi

    # look for a pingback server responsable for the target
    pingserver=`_isPingbackEnabled $pingtargets`
    if [ "$pingserver" == "" ]
    then
        echo "Error: No pingback server found for $pingtargets"
        exit 1
    fi

    # finally, do the ping
    response=`_sendPingback $pingserver $pingsource $pingtargets`
    echo "server response: $response"
}

docu_help () { echo "outputs the manpage of $this"; }
do_help ()
{
    _checkTool man
    realfile=`readlink $thisexec`
    if [ "$realfile" == "" ]
    then
        # assume useage over "xxx/yyy/rdf.sh/rdf.sh help"
        execdir=`dirname $thisexec`
        manpage="$execdir/rdf.1"
    else
        # assume rdf.sh started as link and manpage is in same dir with script
        execdir=`dirname $thisexec`
        scriptdir=`dirname $realfile`
        manpage="$execdir/$scriptdir/rdf.1"
    fi
    # try central manpage first, then try the guessed one
    if [[ "$os" == "darwin" ]]; then
        man rdf 2>/dev/null || man $manpage
    else
        man rdf 2>/dev/null || man -l $manpage
    fi
}

###
# execute the command NOW :-)
###

# rdf.sh uses proper XDG config and cache directories now
if [ "$XDG_CONFIG_HOME" == "" ]
then
    XDG_CONFIG_HOME="$HOME/.config"
fi
if [ "$XDG_CACHE_HOME" == "" ]
then
    XDG_CACHE_HOME="$HOME/.cache"
fi
confdir="$XDG_CONFIG_HOME/rdf.sh"
cachedir="$XDG_CACHE_HOME/rdf.sh"
mkdir -p $confdir
mkdir -p $cachedir
historyfile="$cachedir/resource.history"
prefixcache="$cachedir/prefix.cache"
prefixlocal="$confdir/prefix.local"
touch $prefixlocal
touch $prefixcache

# taken from http://stackoverflow.com/questions/2630812/
commandlist=`typeset -f | grep "do_.*()" | cut -d "_" -f 2 | cut -d " " -f 1 | sort`

# if no command is given, present a basic help screen
if [ "$command" == "" ]
then
    echo "$this is a a multi-tool shell script for doing Semantic Web jobs on the command line."
    echo "Version:  $version"
    echo "Homepage: $home"
    echo ""
    echo "Syntax: $this <command>"
    echo ""
    echo "Available commands are:"
    for cmd in $commandlist
    do
        echo "  $cmd:" `docu_$cmd`
    done
    exit 1
fi

# for generating the autocompletion suggestions automatically
if [ "$command" == "zshcomp" ]
then
    #echo "$commandlist"
    echo "("
    for cmd in $commandlist
    do
        echo $cmd:\"`docu_$cmd`\"
    done
    echo ")"
    exit 1
fi

# now start the sub - command
# taken from http://stackoverflow.com/questions/1007538/
if type do_$command >/dev/null 2>&1
then
    do_$command $*
else
    echo "$this: '$command' is not a rdf command. See '$this help'."
    exit 1
fi

